### DISCUSSION TITLE
Is it possible to trigger a snippet with a custom keymap instead of physically typing the trigger words?

### DISCUSSION THREAD

**Discussion by ziontee113** [L3MON4D3/LuaSnip#389] [Ideas]
First of all I'd like to thank LuaSnip developers for the Amazing Plugin!

I've been wondering for a while is it possible to trigger a snippet with a custom keymap.
For example, in Lua: "local xxx = yyy". It would be much easier to press a key combination like <c-u> instead of physically typing out the trigger to expand the snippet!

Maybe we can implement a function like: `callLuaSnippet("trigger")`

With this type of custom mapping available, we could map any key combinations we want! And reduce our cognitive load!

**Comment by L3MON4D3** [comment]
Oh yes, that's very possible :D

`ls.snip_expand(<some_snippet>)` is used to expand a snippet at your cursor position, you'll need to store the snippets somewhere and address them somehow.
I think something like the following would be pretty flexible (not sure if it works exactly like this, I didn't test it).

First a module that maps key -> snippet.
```lua
-- in lua/keymap_snippets.lua

-- add snippet function-definitions first.
ls.setup_snip_env()

return {
	"snip1" = parse("trigger", "expanded with <C-D>")
}
```

Then a helper-function for the keymaps.
```lua
-- in your config:
local function map_snippet(keys, snippet_key)
	vim.keymap.set("i", keys, function()
		require("luasnip").snip_expand(require("keymap_snippets")[snippet_key])
	end)
end

map_snippet("<C-D>", "snip1")
```

You could also use the keys for accessing the snippets as key-combinations (`{"<C-D>" = parse("trigger", "expanded with <C-D>")}`), might be even cleaner.

**Reply by ziontee113** [reply]
Thank you! This is my current solution:

```lua
-- Autocmd for keymap triggered snippets --
local autocmd = vim.api.nvim_create_autocmd
local augroup = vim.api.nvim_create_augroup
local map = vim.keymap.set
local opts = { noremap = true, silent = true }

local snippets = { a bunch of snippets }
local autosnippets = { a bunch of autosnippets }

local group = augroup("Lua Snippets", { clear = true })
autocmd("BufEnter", {
	pattern = "*.lua",
	group = group,
	callback = function()
		map({ "i" }, "jj", function()
			ls.snip_expand(vim.deepcopy(snippets[1]))
		end, opts)
	end,
})

return snippets, autosnippets
```


https://user-images.githubusercontent.com/102876811/163695495-3e049722-75aa-4042-87e4-12ee17b10b65.mp4



I'm currently using `require("luasnip.loaders.from_lua").load({ paths = "~/.config/nvim/snippets/" })` to hot reload snippets. 🔥
To reload those autocmds I use `:luafile %<cr>`, go to some other buffer then go back.

Is the "BufEnter" event bad for this usecase? Or it it a good one? 🤔

Also do you think it is possible to make the snippets files in the subfolders also hot-reloadable? Currently the main files are hot-reloaded at save. But the ones in the subfolders isn't. 😄

**Reply by L3MON4D3** [reply]
Nice, or like that for snippets that should be regularly expandable too.
I don't really get the `autocmd`; can't you just call `map` directly?
Also, this is one of the few places where you don't need the `deepcopy`, the snippet is copied in `snip_expand` :P

The files in subfolder not being reloadable is surprising and definitely a bug. I'll have a look at it.

**Reply by ziontee113** [reply]
Thank you :) The `autocmd` is for different filetypes.
In my snippets folder I have javascript.lua, lua.lua, all.lua, etc...
In my `*.js` files I can map `jj` to `console.log()` and in `lua` I can have `jj` for `local {} = {}` for example 👍

**Reply by L3MON4D3** [reply]
Ahh ofc, that makes sense.
BufEnter seems pretty good for that usecase, I can't think of a better way👍

**Comment by tjex** [comment]
Hoping I can get some help with this. 

I'm also trying to get this working, but has something perhaps changed in the api since last posted?
I feel like I'm doing everything correct:

~/.config/nvim/lua/tjex/keybind-snips.lua:
```lua
ls.setup_snip_env()

return {
	s(
	"print",
	fmt(
	    [[
                 fmt.Println({})
            ]],
	    { i(1) }
         )
	),
}


```


`~/.config/nvim/after/plugin/luasnip.lua`:
```lua
require("luasnip.loaders.from_lua").load { paths = "~/.config/nvim/snippets/" }

-- other configs

local function map_snippet(keys, snippet_key)
	vim.keymap.set("i", keys, function()
		require("luasnip").snip_expand(require("tjex.keybind-snips")[snippet_key])
	end)
end

map_snippet("<C-k>", "print")
```

But I get this error message:
```text
E5108: Error executing lua: ...nvim/site/pack/packer/start/LuaSnip/lua/luasnip/init.lua:229: attempt to index local 'snippet' (a nil value)
stack traceback:
        ...nvim/site/pack/packer/start/LuaSnip/lua/luasnip/init.lua:229: in function 'snip_expand'
        [string ":source (no file)"]:42: in function 'callback'
        ...site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:159: in function 'callback'
        ...m/site/pack/packer/start/nvim-cmp/lua/cmp/utils/misc.lua:13: in function 'fallback'
        ...te/pack/packer/start/nvim-cmp/lua/cmp/config/mapping.lua:215: in function 'on_keymap'
        ...re/nvim/site/pack/packer/start/nvim-cmp/lua/cmp/core.lua:145: in function 'callback'
        ...site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:133: in function <...site/pack/packer/start/nvim-cmp/lua/cmp/utils/keymap.lua:127>
```

**Reply by L3MON4D3** [reply]
Ah, `print` is nil in the table returned by the module.
Either make it `print = s(...` or call `map_snippet` with 1 instead of "print" (but the first option is better imo)

**Reply by tjex** [reply]
yus! Thanks so much.

Here is my more hacky approach to specify file type specific mappings too, if it helps or gives ideas to others.
`pcall` isn't working though for some reason 🤷‍♂️. Don't get any error when executing on filetype I don't have a `keysnips-<lang>.lua` for. But at least I can name different print functions as "print" in different `keysnips-<lang>.lua`, allowing me to keep the same keybind.

~/.config/nvim/after/plugin/luasnip.lua:

```lua
local function map_snippet(keys, snippet_key)
	vim.keymap.set("i", keys, function()
		local ftype = vim.bo.filetype
		local function keybind_snip()
			require("luasnip").snip_expand(require("snippets.keysnips-" .. ftype)[snippet_key]) -- nvim/lua/keybind-snips
		end
        local snip_found, _ = pcall(keybind_snip)
		if not snip_found then
			print("no keybind snippets for this filetype")
			return
		end
	end)
end
```

~/.config/nvim/lua/snippets:
```tree
.
├── all.lua
├── bash.lua
├── go.lua
├── keysnips-go.lua
├── lua.lua
└── markdown.lua
```

**Reply by emilBeBri** [reply]
YES have been looking for this for some time, your solution works great @tjex , thanks!
